api/agent/slots.go

```diff
@@ -270,31 +272,70 @@ func (a *slotQueueMgr) deleteSlotQueue(slots *slotQueue) bool {
 	return isDeleted
 }
 
+var shapool = &sync.Pool{New: func() interface{} { return sha1.New() }}
+
+// TODO do better; once we have app+route versions this function
+// can be simply app+route names & version
 func getSlotQueueKey(call *call) string {
 	// return a sha1 hash of a (hopefully) unique string of all the config
 	// values, to make map lookups quicker [than the giant unique string]
 
-	hash := sha1.New()
-	fmt.Fprint(hash, call.AppName, "\x00")
-	fmt.Fprint(hash, call.Path, "\x00")
-	fmt.Fprint(hash, call.Image, "\x00")
-	fmt.Fprint(hash, call.Timeout, "\x00")
-	fmt.Fprint(hash, call.IdleTimeout, "\x00")
-	fmt.Fprint(hash, call.Memory, "\x00")
-	fmt.Fprint(hash, call.CPUs, "\x00")
-	fmt.Fprint(hash, call.Format, "\x00")
-
-	// we have to sort these before printing, yay. TODO do better
+	hash := shapool.Get().(hash.Hash)
+	hash.Reset()
+	defer shapool.Put(hash)
+
+	hash.Write(unsafeBytes(call.AppName))
+	hash.Write(unsafeBytes("\x00"))
+	hash.Write(unsafeBytes(call.Path))
+	hash.Write(unsafeBytes("\x00"))
+	hash.Write(unsafeBytes(call.Image))
+	hash.Write(unsafeBytes("\x00"))
+	hash.Write(unsafeBytes(call.Format))
+	hash.Write(unsafeBytes("\x00"))
+
+	// these are all static in size we only need to delimit the whole block of them
+	var byt [8]byte
+	binary.LittleEndian.PutUint32(byt[:4], uint32(call.Timeout))
+	hash.Write(byt[:4])
+
+	binary.LittleEndian.PutUint32(byt[:4], uint32(call.IdleTimeout))
+	hash.Write(byt[:4])
+
+	binary.LittleEndian.PutUint64(byt[:], call.Memory)
+	hash.Write(byt[:])
+
+	binary.LittleEndian.PutUint64(byt[:], uint64(call.CPUs))
+	hash.Write(byt[:])
+	hash.Write(unsafeBytes("\x00"))
+
+	// we have to sort these before printing, yay.
+	// TODO if we had a max size for config const we could avoid this!
 	keys := make([]string, 0, len(call.Config))
 	for k := range call.Config {
-		keys = append(keys, k)
+		i := sort.SearchStrings(keys, k)
```

Comment:
avoids recursive sort and hence faster?

Reply:
oh I see, memmove + bin search insertion sort.

Reply:
this reduces the number of times we do linear operations to 2. previously we ran over call.Config to append keys, then sort, then add the strings. we still have the same asymptotic bounds as the sort (nlogn) [I think] since we're doing logn search n times, but we save one extra n by doing it in this loop we were already taking. it's not much, but it's something (the nature of this patch).